libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -199,187 +231,214 @@ export class ConfigurationFile<TConfigurationFile> {
     if (!cacheEntry) {
-      try {
-        const resolvedConfigurationFilePathForErrors: string = ConfigurationFile._formatPathForError(
-          resolvedConfigurationFilePath
-        );
-
-        if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
-          throw new Error(
-            'A loop has been detected in the "extends" properties of configuration file at ' +
-              `"${resolvedConfigurationFilePathForErrors}".`
-          );
-        }
+      cacheEntry = await this._generateCacheEntryAsync(
+        resolvedConfigurationFilePath,
+        options,
+        visitedConfigurationFilePaths
+      );
+      this._configurationFileCache.set(resolvedConfigurationFilePath, cacheEntry);
+    }
 
-        visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
+    if (cacheEntry.parentFileError) {
+      throw cacheEntry.parentFileError;
+    } else if (cacheEntry.error) {
+      if (!isFirstFile || !options.ignoreIfNotExist || !FileSystem.isNotExistError(cacheEntry.error)) {
+        throw cacheEntry.error;
+      } else {
+        return undefined;
+      }
+    } else {
+      return cacheEntry.configurationFile;
+    }
+  }
 
-        let fileText: string;
-        try {
-          fileText = await FileSystem.readFileAsync(resolvedConfigurationFilePath);
-        } catch (e) {
-          if (FileSystem.isNotExistError(e)) {
-            e.message = `File does not exist: ${resolvedConfigurationFilePathForErrors}`;
-          }
+  private async _generateCacheEntryAsync(
+    resolvedConfigurationFilePath: string,
+    options: ILoadConfigurationFileOptions,
+    visitedConfigurationFilePaths: Set<string>
+  ): Promise<IConfigurationFileCacheEntry<TConfigurationFile>> {
+    try {
+      const resolvedConfigurationFilePathForErrors: string = ConfigurationFile._formatPathForError(
+        resolvedConfigurationFilePath
+      );
+
+      if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
+        throw new Error(
+          'A loop has been detected in the "extends" properties of configuration file at ' +
+            `"${resolvedConfigurationFilePathForErrors}".`
+        );
+      }
 
-          throw e;
-        }
+      visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
 
-        let configurationJson: IConfigurationJson & TConfigurationFile;
-        try {
-          configurationJson = await JsonFile.parseString(fileText);
-        } catch (e) {
-          throw new Error(`In config file "${resolvedConfigurationFilePathForErrors}": ${e}`);
+      let fileText: string;
+      try {
+        fileText = await FileSystem.readFileAsync(resolvedConfigurationFilePath);
+      } catch (e) {
+        if (FileSystem.isNotExistError(e)) {
+          e.message = `File does not exist: ${resolvedConfigurationFilePathForErrors}`;
         }
 
-        this._schema.validateObject(configurationJson, resolvedConfigurationFilePathForErrors);
+        throw e;
```

Comment:
Is this something you would want to set on the `error` property and propagate up instead?

Reply:
Ah, the whole thing is wrapped in a try-catch. Maybe return early for clarity then instead of the large try-catch block?

Reply:
I went back-and-forth on that and I thought this was more readable. @octogonz, thoughts?